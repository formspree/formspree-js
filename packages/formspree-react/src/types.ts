/**
 * ExtraData values can be strings or functions that return a string, or a
 * promise that resolves to a string. Errors should be handled internally.
 * Functions can return undefined to skip this ExtraData value.
 */
export type ExtraDataValue =
  | undefined
  | string
  | (() => string)
  | (() => Promise<string>)
  | (() => undefined)
  | (() => Promise<undefined>);

export type ExtraData = {
  [key: string]: ExtraDataValue;
};
