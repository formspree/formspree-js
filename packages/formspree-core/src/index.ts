export {
  createClient,
  getDefaultClient,
  type Client,
  type Config,
} from './core';

export {
  SubmissionError,
  isSubmissionError,
  type FieldErrorCode,
  type FieldValues,
  type FormErrorCode,
  type SubmissionData,
  type SubmissionOptions,
  type SubmissionResult,
  type SubmissionSuccess,
} from './submission';

export { appendExtraData } from './utils';
