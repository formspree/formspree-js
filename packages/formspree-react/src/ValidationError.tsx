import React from 'react';
import { FormError } from '@formspree/core';

export type ValidationErrorProps = {
  prefix?: string;
  field?: string;
  errors: FormError[];
  [x: string]: any;
}

export const ValidationError: React.FC<ValidationErrorProps> = props => {
  const { prefix, field, errors, ...attrs } = props;

  const error = (errors || []).find(error => {
    return error.field === field;
  });

  if (!error) {
    return null;
  }

  return (
    <div {...attrs}>
      {prefix} {error.message}
    </div>
  );
};
