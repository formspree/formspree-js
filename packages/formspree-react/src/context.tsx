import { createClient, getDefaultClient, type Client } from '@formspree/core';
import { Elements } from '@stripe/react-stripe-js';
import { loadStripe } from '@stripe/stripe-js/pure.js';
import React, {
  useContext,
  useEffect,
  useMemo,
  useState,
  type ReactNode,
} from 'react';

export type FormspreeContextType = {
  client: Client;
};

export type FormspreeProviderProps = {
  children: ReactNode;
  project?: string;
  stripePK?: string;
};

const FormspreeContext = React.createContext<FormspreeContextType | null>(null);

/**
 * FormspreeProvider creates Formspree Client based on the given props
 * and makes the client available via context.
 */
export function FormspreeProvider(props: FormspreeProviderProps) {
  const { children, project, stripePK } = props;
  const [client, setClient] = useState(createClient({ project }));
  const stripePromise = useMemo(
    () => (stripePK ? loadStripe(stripePK) : null),
    [stripePK]
  );

  useEffect(() => {
    let isMounted = true;
    if (isMounted) {
      setClient((client) =>
        client.project !== project
          ? createClient({ ...client, project })
          : client
      );
    }
    return () => {
      isMounted = false;
    };
  }, [project]);

  useEffect(() => {
    let isMounted = true;
    stripePromise?.then((stripe) => {
      if (isMounted && stripe) {
        setClient((client) => createClient({ ...client, stripe }));
      }
    });
    return () => {
      isMounted = false;
    };
  }, [stripePromise]);

  return (
    <FormspreeContext.Provider value={{ client }}>
      {stripePromise ? (
        <Elements stripe={stripePromise}>{children}</Elements>
      ) : (
        children
      )}
    </FormspreeContext.Provider>
  );
}

export function useFormspree(): FormspreeContextType {
  return useContext(FormspreeContext) ?? { client: getDefaultClient() };
}
