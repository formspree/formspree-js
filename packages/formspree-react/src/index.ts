export { CardElement } from '@stripe/react-stripe-js';
export { ValidationError, type ValidationErrorProps } from './ValidationError';
export {
  FormspreeProvider,
  useFormspree,
  type FormspreeContextType,
  type FormspreeProviderProps,
} from './context';
export * from './types';
export { useForm, type TUseForm } from './useForm';
export { useSubmit } from './useSubmit';
