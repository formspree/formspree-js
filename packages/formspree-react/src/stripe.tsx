import React from 'react';
import { useElements } from '@stripe/react-stripe-js';
import type { StripeElements } from '@stripe/stripe-js';
import { createContext, useContext, type PropsWithChildren } from 'react';

type StripeContextValue = {
  elements: StripeElements | null;
};

const StripeContext = createContext<StripeContextValue>({ elements: null });

export function StripeProvider(props: PropsWithChildren) {
  const { children } = props;
  const elements = useElements();
  return (
    <StripeContext.Provider value={{ elements }}>
      {children}
    </StripeContext.Provider>
  );
}

export function useStripeContext(): StripeContextValue {
  return useContext(StripeContext);
}
