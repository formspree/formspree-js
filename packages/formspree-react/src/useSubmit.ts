import {
  appendExtraData,
  isSubmissionError,
  type Client,
  type FieldValues,
  type SubmissionData,
  type SubmissionError,
  type SubmissionSuccess,
} from '@formspree/core';
import { CardElement } from '@stripe/react-stripe-js';
import type { PaymentMethodResult } from '@stripe/stripe-js';
import { useMemo } from 'react';
import { version } from '../package.json';
import { useFormspree } from './context';
import type { ExtraData } from './types';

const clientName = `@formspree/react@${version}`;

type Options<T extends FieldValues> = {
  client?: Client;
  extraData?: ExtraData;
  onError?: (error: SubmissionError<T>) => void;
  onSuccess?: (data: SubmissionSuccess) => void;
  // origin overrides the submission origin (default: "https://formspree.io")
  origin?: string;
};

type FormEvent = React.FormEvent<HTMLFormElement>;

export type SubmitHandler<T extends FieldValues> = (
  submission: FormEvent | SubmissionData<T>
) => Promise<void>;

export function useSubmit<T extends FieldValues>(
  formKey: string,
  options: Options<T> = {}
): SubmitHandler<T> {
  const formspree = useFormspree();
  const {
    client = formspree.client,
    extraData,
    onError,
    onSuccess,
    origin,
  } = options;

  const { stripe } = client;
  const cardElement = useMemo(
    () => stripe?.elements().getElement(CardElement),
    [stripe]
  );

  return async function handleSubmit(submission) {
    const data = isEvent(submission) ? getFormData(submission) : submission;

    // Append extra data from config
    if (typeof extraData === 'object') {
      for (const [prop, value] of Object.entries(extraData)) {
        let extraDataValue: string | undefined;
        if (typeof value === 'function') {
          extraDataValue = await value();
        } else {
          extraDataValue = value;
        }
        if (extraDataValue !== undefined) {
          appendExtraData(data, prop, extraDataValue);
        }
      }
    }

    const result = await client.submitForm(formKey, data, {
      endpoint: origin,
      clientName,
      createPaymentMethod:
        stripe && cardElement
          ? (): Promise<PaymentMethodResult> =>
              stripe.createPaymentMethod({
                type: 'card',
                card: cardElement,
                billing_details: mapBillingDetailsPayload(data),
              })
          : undefined,
    });

    if (isSubmissionError(result)) {
      onError?.(result);
    } else {
      onSuccess?.(result);
    }
  };
}

function isEvent(
  submission: FormEvent | SubmissionData
): submission is FormEvent {
  return (
    'preventDefault' in submission &&
    typeof submission.preventDefault === 'function'
  );
}

function getFormData(event: FormEvent): FormData {
  event.preventDefault();
  const form = event.currentTarget;
  if (form.tagName != 'FORM') {
    throw new Error('submit was triggered for a non-form element');
  }
  return new FormData(form);
}

type StripeBillingDetailsPayload = {
  address: StripeAddressPayload;
  name?: string;
  email?: string;
  phone?: string;
};

function mapBillingDetailsPayload(
  data: SubmissionData
): StripeBillingDetailsPayload {
  const billing: StripeBillingDetailsPayload = {
    address: mapAddressPayload(data),
  };

  for (const key of ['name', 'email', 'phone'] as const) {
    const value = data instanceof FormData ? data.get(key) : data[key];
    if (value && typeof value === 'string') {
      billing[key] = value;
    }
  }

  return billing;
}

type StripeAddressPayload = Partial<{
  line1: string;
  line2: string;
  city: string;
  country: string;
  state: string;
  postal_code: string;
}>;

function mapAddressPayload(data: SubmissionData): StripeAddressPayload {
  const address: StripeAddressPayload = {};

  for (const [fromKey, toKey] of [
    ['address_line1', 'line1'],
    ['address_line2', 'line2'],
    ['address_city', 'city'],
    ['address_country', 'country'],
    ['address_state', 'state'],
    ['address_postal_code', 'postal_code'],
  ] as const) {
    const value = data instanceof FormData ? data.get(fromKey) : data[fromKey];
    if (value && typeof value === 'string') {
      address[toKey] = value;
    }
  }

  return address;
}
