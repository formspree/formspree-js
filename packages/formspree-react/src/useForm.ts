import {
  type Client,
  type FieldValues,
  type SubmissionError,
  type SubmissionSuccess,
} from '@formspree/core';
import { useState } from 'react';
import type { ExtraData } from './types';
import { useSubmit, type SubmitHandler } from './useSubmit';

type ResetFunction = () => void;

export type TUseForm<T extends FieldValues> = [
  {
    errors: SubmissionError<T> | null;
    result: SubmissionSuccess | null;
    submitting: boolean;
    succeeded: boolean;
  },
  SubmitHandler<T>,
  ResetFunction
];

export function useForm<T extends FieldValues>(
  formKey: string,
  args: {
    client?: Client;
    data?: ExtraData;
    endpoint?: string;
  } = {}
): TUseForm<T> {
  const [errors, setErrors] = useState<SubmissionError<T> | null>(null);
  const [result, setResult] = useState<SubmissionSuccess | null>(null);
  const [submitting, setSubmitting] = useState(false);
  const [succeeded, setSucceeded] = useState(false);

  if (!formKey) {
    throw new Error(
      'You must provide a form key or hashid ' +
        '(e.g. useForm("myForm") or useForm("123xyz")'
    );
  }

  const submit = useSubmit<T>(formKey, {
    client: args.client,
    extraData: args.data,
    onError(error) {
      setErrors(error);
      setSubmitting(false);
      setSucceeded(false);
    },
    onSuccess(result) {
      setErrors(null);
      setResult(result);
      setSubmitting(false);
      setSucceeded(true);
    },
    origin: args.endpoint,
  });

  return [
    { errors, result, submitting, succeeded },

    async function handleSubmit(submissionData) {
      setSubmitting(true);
      await submit(submissionData);
    },

    function reset() {
      setErrors(null);
      setResult(null);
      setSubmitting(false);
      setSucceeded(false);
    },
  ];
}
