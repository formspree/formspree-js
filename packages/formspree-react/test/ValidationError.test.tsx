import React from 'react';
import ReactDOM from 'react-dom';
import ReactTestUtils from 'react-dom/test-utils';
import { ValidationError } from '../src';

const { act } = ReactTestUtils;

let container: HTMLElement;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

it('renders a field error if one exists', () => {
  act(() => {
    ReactDOM.render(
      <ValidationError
        prefix="Email"
        field="email"
        className="error"
        errors={[
          {
            field: 'email',
            message: 'is required',
            code: 'REQUIRED'
          }
        ]}
      />,
      container
    );
  });

  expect(container).toMatchSnapshot();
});

it('renders field-less errors', () => {
  act(() => {
    ReactDOM.render(
      <ValidationError
        className="error"
        errors={[
          {
            message: 'Form is disabled',
            code: 'FORM_DISABLED'
          }
        ]}
      />,
      container
    );
  });

  expect(container).toMatchSnapshot();
});

it('does not render anything if the field does not have an error', () => {
  act(() => {
    ReactDOM.render(
      <ValidationError
        prefix="Email"
        field="email"
        errors={[
          {
            field: 'name',
            message: 'is required',
            code: 'REQUIRED'
          }
        ]}
      />,
      container
    );
  });

  expect(container).toMatchSnapshot();
});
