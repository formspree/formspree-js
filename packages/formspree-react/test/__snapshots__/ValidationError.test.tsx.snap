// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render anything if the field does not have an error 1`] = `<div />`;

exports[`renders a field error if one exists 1`] = `
<div>
  <div
    class="error"
  >
    Email
     
    is required
  </div>
</div>
`;

exports[`renders field-less errors 1`] = `
<div>
  <div
    class="error"
  >
     
    Form is disabled
  </div>
</div>
`;
